File: dqops/src/main/java/com/dqops/rest/models/metadata/ColumnStatisticsModel.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */
package com.dqops.rest.models.metadata;

import com.dqops.data.statistics.models.StatisticsMetricModel;
import com.dqops.data.statistics.models.StatisticsResultsForColumnModel;
import com.dqops.metadata.search.StatisticsCollectorSearchFilters;
import com.dqops.metadata.sources.ColumnSpec;
import com.dqops.metadata.sources.ColumnTypeSnapshotSpec;
import com.dqops.metadata.sources.PhysicalTableName;
import com.dqops.utils.docs.generators.SampleStringsRegistry;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

import java.util.Collection;

/**
 * Column model that returns the basic fields from a column specification and a summary of the most recent statistics collection.
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "ColumnStatisticsModel", description = "Column model that returns the basic fields from a column specification with the additional data statistics.")
public class ColumnStatisticsModel {
    @JsonPropertyDescription("Connection name.")
    private String connectionName;

    @JsonPropertyDescription("Physical table name including the schema and table names.")
    private PhysicalTableName table;

    @JsonPropertyDescription("Column name.")
    private String columnName;

    @JsonPropertyDescription("Column hash that identifies the column using a unique hash code.")
    private Long columnHash;

    @JsonPropertyDescription("Disables all data quality checks on the column. Data quality checks will not be executed.")
    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    private boolean disabled;

    @JsonPropertyDescription("True when the column has any checks configured.")
    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    private boolean hasAnyConfiguredChecks;

    @JsonPropertyDescription("Column data type that was retrieved when the table metadata was imported.")
    @JsonInclude(JsonInclude.Include.NON_NULL)
    private ColumnTypeSnapshotSpec typeSnapshot;

    @JsonPropertyDescription("List of collected column statistics.")
    @JsonInclude(JsonInclude.Include.NON_NULL)
    private Collection<StatisticsMetricModel> statistics;

    /**
     * Configured parameters for the "collect statistics" job that should be pushed to the job queue in order to run all statistics collectors for this column.
     */
    @JsonPropertyDescription("Configured parameters for the \"collect statistics\" job that should be pushed to the job queue in order to run all statistics collectors for this column")
    private StatisticsCollectorSearchFilters collectColumnStatisticsJobTemplate;

    /**
     * Boolean flag that decides if the current user can collect statistics.
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can collect statistics.")
    private boolean canCollectStatistics;

    /**
     * Creates a column profile model from a column specification for a requested profile.
     * This model is used on a profiler summary screen.
     * @param physicalTableName Physical table name.
     * @param columnName        Column name.
     * @param columnSpec        Source column specification.
     * @param statisticsResultsForColumn List of column metrics.
     * @param canRunStatisticsJob The current user can collect statistics.
     * @return Column statistics model.
     */
    public static ColumnStatisticsModel fromColumnSpecificationAndStatistic(String connectionName,
                                                                            PhysicalTableName physicalTableName,
                                                                            String columnName,
                                                                            ColumnSpec columnSpec,
                                                                            StatisticsResultsForColumnModel statisticsResultsForColumn,
                                                                            boolean canRunStatisticsJob) {
        return new ColumnStatisticsModel() {{
            setConnectionName(connectionName);
            setColumnHash(columnSpec.getHierarchyId() != null ? columnSpec.getHierarchyId().hashCode64() : null);
            setTable(physicalTableName);
            setColumnName(columnName);
            setDisabled(columnSpec.isDisabled());
            setTypeSnapshot(columnSpec.getTypeSnapshot());
            setHasAnyConfiguredChecks(columnSpec.hasAnyChecksConfigured());
            setCanCollectStatistics(canRunStatisticsJob);
            setStatistics(statisticsResultsForColumn != null ? statisticsResultsForColumn.getMetrics() : null);
            setCollectColumnStatisticsJobTemplate(new StatisticsCollectorSearchFilters()
            {{
                setConnection(connectionName);
                setFullTableName(physicalTableName.toTableSearchFilter());
                getColumnNames().add(columnName);
                setEnabled(true);
            }});
        }};
    }

    public static class ColumnStatisticsModelSampleFactory implements SampleValueFactory<ColumnStatisticsModel> {
        @Override
        public ColumnStatisticsModel createSample() {
            return ColumnStatisticsModel.fromColumnSpecificationAndStatistic(
                    SampleStringsRegistry.getConnectionName(),
                    PhysicalTableName.fromSchemaTableFilter(SampleStringsRegistry.getSchemaTableName()),
                    SampleStringsRegistry.getColumnName(),
                     new ColumnSpec.ColumnSpecSampleFactory().createSample(),
                    new StatisticsResultsForColumnModel.StatisticsResultsForColumnModelSampleFactory().createSample(),
                    true
            );
        }
    }
}

File: distribution/python/dqops/client/models/column_statistics_model.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.column_type_snapshot_spec import ColumnTypeSnapshotSpec
    from ..models.physical_table_name import PhysicalTableName
    from ..models.statistics_collector_search_filters import (
        StatisticsCollectorSearchFilters,
    )
    from ..models.statistics_metric_model import StatisticsMetricModel


T = TypeVar("T", bound="ColumnStatisticsModel")


@_attrs_define
class ColumnStatisticsModel:
    """Column model that returns the basic fields from a column specification with the additional data statistics.

    Attributes:
        connection_name (Union[Unset, str]): Connection name.
        table (Union[Unset, PhysicalTableName]):
        column_name (Union[Unset, str]): Column name.
        column_hash (Union[Unset, int]): Column hash that identifies the column using a unique hash code.
        disabled (Union[Unset, bool]): Disables all data quality checks on the column. Data quality checks will not be
            executed.
        has_any_configured_checks (Union[Unset, bool]): True when the column has any checks configured.
        type_snapshot (Union[Unset, ColumnTypeSnapshotSpec]):
        statistics (Union[Unset, List['StatisticsMetricModel']]): List of collected column statistics.
        collect_column_statistics_job_template (Union[Unset, StatisticsCollectorSearchFilters]):
        can_collect_statistics (Union[Unset, bool]): Boolean flag that decides if the current user can collect
            statistics.
    """

    connection_name: Union[Unset, str] = UNSET
    table: Union[Unset, "PhysicalTableName"] = UNSET
    column_name: Union[Unset, str] = UNSET
    column_hash: Union[Unset, int] = UNSET
    disabled: Union[Unset, bool] = UNSET
    has_any_configured_checks: Union[Unset, bool] = UNSET
    type_snapshot: Union[Unset, "ColumnTypeSnapshotSpec"] = UNSET
    statistics: Union[Unset, List["StatisticsMetricModel"]] = UNSET
    collect_column_statistics_job_template: Union[
        Unset, "StatisticsCollectorSearchFilters"
    ] = UNSET
    can_collect_statistics: Union[Unset, bool] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        connection_name = self.connection_name
        table: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.table, Unset):
            table = self.table.to_dict()

        column_name = self.column_name
        column_hash = self.column_hash
        disabled = self.disabled
        has_any_configured_checks = self.has_any_configured_checks
        type_snapshot: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.type_snapshot, Unset):
            type_snapshot = self.type_snapshot.to_dict()

        statistics: Union[Unset, List[Dict[str, Any]]] = UNSET
        if not isinstance(self.statistics, Unset):
            statistics = []
            for statistics_item_data in self.statistics:
                statistics_item = statistics_item_data.to_dict()

                statistics.append(statistics_item)

        collect_column_statistics_job_template: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.collect_column_statistics_job_template, Unset):
            collect_column_statistics_job_template = (
                self.collect_column_statistics_job_template.to_dict()
            )

        can_collect_statistics = self.can_collect_statistics

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if connection_name is not UNSET:
            field_dict["connection_name"] = connection_name
        if table is not UNSET:
            field_dict["table"] = table
        if column_name is not UNSET:
            field_dict["column_name"] = column_name
        if column_hash is not UNSET:
            field_dict["column_hash"] = column_hash
        if disabled is not UNSET:
            field_dict["disabled"] = disabled
        if has_any_configured_checks is not UNSET:
            field_dict["has_any_configured_checks"] = has_any_configured_checks
        if type_snapshot is not UNSET:
            field_dict["type_snapshot"] = type_snapshot
        if statistics is not UNSET:
            field_dict["statistics"] = statistics
        if collect_column_statistics_job_template is not UNSET:
            field_dict["collect_column_statistics_job_template"] = (
                collect_column_statistics_job_template
            )
        if can_collect_statistics is not UNSET:
            field_dict["can_collect_statistics"] = can_collect_statistics

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.column_type_snapshot_spec import ColumnTypeSnapshotSpec
        from ..models.physical_table_name import PhysicalTableName
        from ..models.statistics_collector_search_filters import (
            StatisticsCollectorSearchFilters,
        )
        from ..models.statistics_metric_model import StatisticsMetricModel

        d = src_dict.copy()
        connection_name = d.pop("connection_name", UNSET)

        _table = d.pop("table", UNSET)
        table: Union[Unset, PhysicalTableName]
        if isinstance(_table, Unset):
            table = UNSET
        else:
            table = PhysicalTableName.from_dict(_table)

        column_name = d.pop("column_name", UNSET)

        column_hash = d.pop("column_hash", UNSET)

        disabled = d.pop("disabled", UNSET)

        has_any_configured_checks = d.pop("has_any_configured_checks", UNSET)

        _type_snapshot = d.pop("type_snapshot", UNSET)
        type_snapshot: Union[Unset, ColumnTypeSnapshotSpec]
        if isinstance(_type_snapshot, Unset):
            type_snapshot = UNSET
        else:
            type_snapshot = ColumnTypeSnapshotSpec.from_dict(_type_snapshot)

        statistics = []
        _statistics = d.pop("statistics", UNSET)
        for statistics_item_data in _statistics or []:
            statistics_item = StatisticsMetricModel.from_dict(statistics_item_data)

            statistics.append(statistics_item)

        _collect_column_statistics_job_template = d.pop(
            "collect_column_statistics_job_template", UNSET
        )
        collect_column_statistics_job_template: Union[
            Unset, StatisticsCollectorSearchFilters
        ]
        if isinstance(_collect_column_statistics_job_template, Unset):
            collect_column_statistics_job_template = UNSET
        else:
            collect_column_statistics_job_template = (
                StatisticsCollectorSearchFilters.from_dict(
                    _collect_column_statistics_job_template
                )
            )

        can_collect_statistics = d.pop("can_collect_statistics", UNSET)

        column_statistics_model = cls(
            connection_name=connection_name,
            table=table,
            column_name=column_name,
            column_hash=column_hash,
            disabled=disabled,
            has_any_configured_checks=has_any_configured_checks,
            type_snapshot=type_snapshot,
            statistics=statistics,
            collect_column_statistics_job_template=collect_column_statistics_job_template,
            can_collect_statistics=can_collect_statistics,
        )

        column_statistics_model.additional_properties = d
        return column_statistics_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
